infobip-jackson-extension-api/src/main/java/com/infobip/jackson/PresentPropertyJsonTypeResolver.java

```diff
@@ -14,11 +14,33 @@ public PresentPropertyJsonTypeResolver(Class<E> type) {
     @Override
     public final Class<?> resolve(Map<String, Object> json) {
         for (E constant : type.getEnumConstants()) {
-            if (json.containsKey(constant.toString().toLowerCase())) {
+            if (json.containsKey(resolvePropertyName(constant))) {
                 return constant.getType();
             }
         }
 
         throw new IllegalArgumentException("Failed to resolve type " + json);
     }
+
+    private String resolvePropertyName(E constant) {
+        for (Type type : getAllInterfaces(constant.getClass())) {
+            if (type.equals(NamedPropertyTypeProvider.class)) {
+                return ((NamedPropertyTypeProvider) constant).getJsonPropertyName();
+            }
+        }
+        return constant.toString().toLowerCase();
+    }
+
+    private List<Type> getAllInterfaces(Class<?> type) {
```

Comment:
This method is copied from [JsonTypeResolverFactory](https://github.com/infobip/infobip-jackson-extension/blob/2ef97a40b7ebb2ce98f2a96568fef6e65bae7fd5/infobip-jackson-extension-module/src/main/java/com/infobip/jackson/JsonTypeResolverFactory.java#L102). Maybe it should extracted to some util class? Or maybe assume that the `NamedPropertyTypeProvider` have to be implemented directly on enum - then, instead of this method, calling `constant.getClass().getInterfaces()` would be enough. What do you think?